app/backend/src/couchers/db.py

```diff
@@ -118,19 +118,39 @@ def new_password_reset_token(session, user, hours=2):
     return password_reset_token, f"{hours} hours"
 
 
-def set_email_change_token(session, user, hours=2):
+def set_email_change_tokens(user, confirm_with_both_emails, hours=2):
     """
-    Make a new email change token that's valid for `hours` hours for this user
+    If the user does not have a password, they need to confirm their email change via their old email in addition to their new email; 'confirm_with_both_emails' flags if confirmation via the old email is required
+
+    Make email change tokens which are valid for `hours` hours
 
     Note: does not call session.commit()
 
-    Returns token and expiry text
+    Returns two tokens and expiry text
     """
-    token = urlsafe_secure_token()
-    user.new_email_token = token
+    old_email_token = ""
+    if confirm_with_both_emails:
+        old_email_token = urlsafe_secure_token()
+        user.old_email_token = old_email_token
+        user.old_email_token_created = now()
+        user.old_email_token_expiry = now() + timedelta(hours=hours)
+        user.need_to_confirm_via_old_email = True
+    else:
+        user.old_email_token = None
+        user.old_email_token_created = None
+        user.old_email_token_expiry = None
+        user.need_to_confirm_via_old_email = False
+
+    new_email_token = urlsafe_secure_token()
+    user.new_email_token = new_email_token
     user.new_email_token_created = now()
     user.new_email_token_expiry = now() + timedelta(hours=hours)
-    return token, f"{hours} hours"
+    user.need_to_confirm_via_new_email = True
+
+    # Check for edge case, however remote
+    if old_email_token == new_email_token:
+        return set_email_change_tokens(user, confirm_with_both_emails)
```

Comment:
Why would `old_email_token == new_email_token`?

Reply:
If the two URLs happen to be the same. Odds are close to zero, but if we ever scale up to millions of users, it's still close to zero but slightly larger.

Reply:
No need to test for that. It's practically impossible for that to happen (as a mathematician I'm not allowed to call it impossible, but it's basically impossible). Each token has 32 bytes (256 bits), of randomness. So the chance of collision is 1 in 2^512, which is way more than there are atoms in the universe!